master/internal/api_experiment.go

```diff
@@ -289,23 +285,109 @@ func (a *apiServer) GetExperiments(
 	default:
 		orderExpr = fmt.Sprintf("id %s", sortByMap[req.OrderBy])
 	}
+	query = query.OrderExpr(orderExpr)
+
+	// Filtering
+	if req.Description != "" {
+		query = query.Where("e.config->>'description' ILIKE ('%%' || ? || '%%')", req.Description)
+	}
+	if req.Name != "" {
+		query = query.Where("e.config->>'name' ILIKE ('%%' || ? || '%%')", req.Name)
+	}
+	if len(req.Labels) > 0 {
+		// In the event labels were removed, if all were removed we insert null,
+		// which previously broke this query.
+		query = query.Where(`string_to_array(?, ',') <@ ARRAY(SELECT jsonb_array_elements_text(
+				CASE WHEN e.config->'labels'::text = 'null'
+				THEN NULL
+				ELSE e.config->'labels' END
+			))`, strings.Join(req.Labels, ",")) // Trying bun.In doesn't work.
+	}
+	if req.Archived != nil {
+		query = query.Where("e.archived = ?", req.Archived.Value)
+	}
+	if len(req.States) > 0 {
+		var allStates []string
+		for _, state := range req.States {
+			allStates = append(allStates, strings.TrimPrefix(state.String(), "STATE_"))
+		}
+		query = query.Where("e.state IN (?)", bun.In(allStates))
+	}
+	if len(req.Users) > 0 {
+		query = query.Where("u.username IN (?)", bun.In(req.Users))
+	}
+	if len(req.UserIds) > 0 {
+		query = query.Where("e.owner_id IN (?)", bun.In(req.UserIds))
+	}
+	if req.ProjectId != 0 {
+		query = query.Where("e.project_id = ?", req.ProjectId)
+	}
 
 	resp := &apiv1.GetExperimentsResponse{}
-	return resp, a.m.db.QueryProtof(
-		"get_experiments",
-		[]interface{}{orderExpr},
-		resp,
-		stateFilterExpr,
-		archivedExpr,
-		userFilterExpr,
-		userIDFilterExpr,
-		labelFilterExpr,
-		req.Description,
-		req.Name,
-		req.ProjectId,
-		req.Offset,
-		req.Limit,
-	)
+	var err error
+	resp.Pagination, err = experimentsPagination(query, int(req.Offset), int(req.Limit))
+	if err != nil {
+		return nil, err
+	}
+
+	// Bun bug? Still getting results when Limit == 0.
+	if resp.Pagination.EndIndex-resp.Pagination.StartIndex != 0 {
+		if err = query.Scan(ctx); err != nil {
+			return nil, err
+		}
+	}
+
+	resp.Experiments = make([]*experimentv1.Experiment, len(experiments))
+	for i, e := range experiments {
+		if resp.Experiments[i], err = e.ToProto(); err != nil {
+			return nil, err
+		}
+	}
+
+	return resp, nil
+}
+
+func experimentsPagination(query *bun.SelectQuery, offset, limit int) (*apiv1.Pagination, error) {
+	// Count number of items without any limits or offsets.
+	total, err := query.Count(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+
+	// Calculate end and start indexes.
+	startIndex := offset
+	if offset > total || offset < -total {
+		startIndex = total
+	} else if offset < 0 {
+		startIndex = total + offset
+	}
+
+	endIndex := startIndex + limit
+	switch {
+	case limit == -2:
+		endIndex = startIndex
```

Comment:
where is this thing implemented in the current code? why do we even need it?

Reply:
thanks 👍 
it uses it to get the number of active experiments per cluster https://github.com/determined-ai/determined/blob/master/webui/react/src/hooks/useFetch.ts#L25